README.md
# Google Clone Coding Challenge

The project is a simple clone of the google home and results pages.

## Specification

-   Your version of Google should have a homepage, with a logo, search bar and two buttons.
-   Upon pressing one button an API call should be executed and users should be shown a page with ten results.
-   Upon pressing the other, users should be taken to the page of one of the results.

## Installation & Usage

### Installation

-   clone or download the repo.
-   open terminal and navigate to `server` folder.
-   Run `npm install` to install server dependecies.
-   open terminal and navigate to `client` folder.
-   Run `npm install` to install client dependecies.


### Usage

-   Run `npm start` to launch the server.
-   Run `npm start` to launch the client.


## Technologies

### Frontend

-   HTML
-   CSS
-   JavaScript

### Backend

-   Javascript / NodeJs
-   NPM
    -   Express
    -   Cors
    -   Morgan
    -   Google-it
    -   Nodemon
    -   Jest
    -   Supertest

## API Endpoints

-   GET `/`
    -   returns `Hello World!`
-   GET `/search`
    -   requires `search` query parameter
    -   responsds with JSON:

```js
{
	results: [
		{
			title: 'string title',
			link: 'string href',
			snippet: 'string snippet',
		}, // ...
	];
}
```

## Wins & Challenges

### Wins

-   Finding Google-it simplified the back-end

### Challenges

## Licence

none


changelog.md
# Changelog

## Feature plan

### Frontend

-   pages:
    -   Homepage
        -   Logo
        -   Search bar
        -   Search button
        -   I'm feeling lucky
        -   nav bar
        -   footer
        -   search suggestion feature
    -   Results Page
        -   header
            -   logo
            -   search bar (filled with search request)
        -   10 results
        -   related searches
        -   search info box (right hand side)

### Backend

api endpoints:

-   GET /
    -   return "hello world"
-   GET /search
    -   requires a query string: search
    -   fetch from some search api
    -   process search results
    -   return processed results

## TODO

### Frontend

    [X]  create basic homepage
    [X]  create form for search bar
    [X]  fetch search results from backend
    [X]  stub results page
    []  impliment search buttons functionality
    [X]  style footer
    []  add non dead links to homepage
    []  style results page

### Backend

    [X]  create basic api
    [X]  create / endpoint
    [X]  stub /search endpoint
    []   write more test

### Misc

    [X]  write readme
    []  update changelog

## changes

### Frontend

### Backend

-   create initial files, dependencies.
-   create GET `/` end point
-   create GET `/search` using `express.Router()`
-   create `search.js` model using `npm google-it` module


client/index.js
let form = document.querySelector("form");

form.addEventListener("submit", redirectToResults);

function redirectToResults(e) {
  e.preventDefault(e);
  const query = e.target[0].value;
  const currentUrl = window.location.href;
  const newUrl = `${currentUrl}resultPage.html???${query}`;
  window.location.href = newUrl;
}

let googleSearch = document.getElementById("googleSearch");

googleSearch.addEventListener("click", handleGoogleSearchButtonClick);

function handleGoogleSearchButtonClick() {
  let input = document.querySelector("input");
  let query = input.value;
  const currentUrl = window.location.href;
  const newUrl = `${currentUrl}resultPage.html???${query}`;
  window.location.href = newUrl;
}

let imFeelingLuckyButton = document.getElementById("imFeelingLuckyButton");

imFeelingLuckyButton.addEventListener("click", imFeelingLucky);

async function imFeelingLucky() {
  let input = document.querySelector("input");
  let query = input.value;
  let fetchedData = await fetchData(query);
  let link = fetchedData.results[0].link;
  window.location.href = link;
}

function createUrlQuery(query) {
  const url = `http://localhost:5000/search?search=${query}`;
  return url;
}

async function fetchData(query) {
  const url = createUrlQuery(query);
  const fetchResults = await fetch(url)
    .then((res) => res.json())
    .catch(console.error);
  return fetchResults;
}

let searchButton = document.getElementById("searchButton");

searchButton.addEventListener("click", handleGoogleSearchButtonClick);
function handleGoogleSearchButtonClick() {
  let input = document.querySelector("input");
  let query = input.value;
  const currentUrl = window.location.href;
  const newUrl = `http://localhost:3000/resultPage.html???${query}`;
  window.location.href = newUrl;
}


client/results.js
function createUrlQuery() {
  const query = window.location.href.split("???")[1];
  const url = `http://localhost:5000/search?search=${query}`;
  return url;
}

async function fetchData() {
  const url = createUrlQuery();
  const fetchResults = await fetch(url)
    .then((res) => res.json())
    .catch(console.error);
  console.log(fetchResults);
  return fetchResults;
}

function createResult({ link, snippet, title }) {
  const div = document.createElement("div");
  const linkA = document.createElement("a");
  linkA.href = link;
  linkA.textContent = link;
  const titleA = document.createElement("a");
  titleA.href = link;
  titleA.textContent = title;
  const snippetP = document.createElement("p");
  snippetP.textContent = snippet;

  div.appendChild(linkA);
  div.appendChild(titleA);
  div.appendChild(snippetP);
  return div;
}

async function appendResultsToMain() {
  const main = document.querySelector("main");
  const searchResults = await fetchData();
  searchResults.results.forEach((result) => {
    const divToAppend = createResult(result);
    main.appendChild(divToAppend);
  });
}

appendResultsToMain();

let form = document.querySelector("form");

form.addEventListener("submit", redirectToResults);

function redirectToResults(e) {
  e.preventDefault(e);
  const query = e.target[0].value;
  const newUrl = `http://localhost:3000/resultPage.html???${query}`;
  window.location.href = newUrl;
}

let searchButton = document.getElementById("searchButton");

searchButton.addEventListener("click", handleGoogleSearchButtonClick);
function handleGoogleSearchButtonClick() {
  let input = document.querySelector("input");
  let query = input.value;
  const currentUrl = window.location.href;
  const newUrl = `http://localhost:3000/resultPage.html???${query}`;
  window.location.href = newUrl;
}

let homeButton = document.getElementById("googleLogo");

homeButton.addEventListener("click", goHome);

function goHome() {
  window.location.href = "http://localhost:3000/";
}
